File: 234-PalindromeLinkedList/234-PalindromeLinkedList.java
// Last updated: 7/28/2025, 12:30:57 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head==null ||head.next==null) return true;
         ListNode fast =head ,slow =head;
         while(fast.next!=null && fast.next.next!=null)
         {
             fast= fast.next.next;
             slow = slow.next;
         }
         slow.next=rev(slow.next);
         slow= slow.next;
         while(slow!=null)
         {
             if(head.val!=slow.val) return false;
            head=head.next;
            slow=slow.next;
         }
         return true;

    }
    ListNode rev(ListNode head)
    {
        ListNode prev= null , curr=head , nextNode=null;
        while(curr!=null)
        {
            nextNode = curr.next;
            curr.next= prev;
            prev=curr;
            curr=nextNode;
        }
        return prev;
    }
}

File: 234-PalindromeLinkedList/234-PalindromeLinkedList.cpp
// Last updated: 7/28/2025, 12:30:52 PM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseLL(ListNode* head){
        if(head == nullptr || head->next == nullptr) return head;

        ListNode* newHead = reverseLL(head->next);
        ListNode* front = head->next;
        front->next = head;
        head->next = nullptr;
        return newHead;
    }
    bool isPalindrome(ListNode* head) {
        if(head == nullptr || head->next == nullptr) return true;
        ListNode* slow = head ;
        ListNode* fast = head ;

        while(fast->next != nullptr && fast->next->next != nullptr){
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* newHead = reverseLL(slow->next);

        ListNode* first = head ;
        ListNode* sec = newHead ;

        while(sec != nullptr){
            if(first->val != sec->val) {
     
                return false;
            }

            first = first->next;
            sec = sec->next;
        }
      
        return true;
    }
};